jina/peapods/runtimes/__init__.py

```diff
@@ -0,0 +1,186 @@
+        if any(bools):
+            or_event.set()
+        else:
+            or_event.clear()
+
+    for e in events:
+        orify(e, changed)
+    changed()
+    return or_event
+
+
+class RuntimeMeta(type):
+    """Meta class of :class:`BaseRuntime` to enable switching between ``thread`` and ``process`` backend. """
+    _dct = {}
+
+    def __new__(cls, name, bases, dct):
+        _cls = super().__new__(cls, name, bases, dct)
+        RuntimeMeta._dct.update({name: {'cls': cls,
+                                        'name': name,
+                                        'bases': bases,
+                                        'dct': dct}})
+        return _cls
+
+    def __call__(cls, *args, **kwargs) -> 'RuntimeMeta':
+        # switch to the new backend
+        _cls = {
+            'thread': threading.Thread,
+            'process': multiprocessing.Process,
+        }.get(getattr(args[0], 'runtime', 'thread'))
+
+        # rebuild the class according to mro
+        for c in cls.mro()[-2::-1]:
+            arg_cls = RuntimeMeta._dct[c.__name__]['cls']
+            arg_name = RuntimeMeta._dct[c.__name__]['name']
+            arg_dct = RuntimeMeta._dct[c.__name__]['dct']
+            _cls = super().__new__(arg_cls, arg_name, (_cls,), arg_dct)
+
+        return type.__call__(_cls, *args, **kwargs)
+
+
+class BaseRuntime(metaclass=RuntimeMeta):
+    def __init__(self, args: Union['argparse.Namespace', Dict]):
+        super().__init__()
+        self.args = args
+        self.name = self.__class__.__name__  #: this is the process name
+
+        self.is_ready_event = _get_event(self)
+        self.is_shutdown = _get_event(self)
+        self.ready_or_shutdown = _make_or_event(self, self.is_ready_event, self.is_shutdown)
+        self.is_shutdown.clear()
+
+        if 'daemon' in args:
+            self.daemon = args.daemon
+        if 'name' in self.args and self.args.name:
+            self.name = f'runtime-{self.args.name}'
+        if 'role' in self.args and self.args.role == PeaRoleType.PARALLEL:
+            self.name = f'runtime-{self.args.name}-{self.args.pea_id}'
+        if 'role' in self.args and self.args.role == PeaRoleType.HEAD:
+            self.name = f'runtime-{self.args.name}-head'
+        if 'role' in self.args and self.args.role == PeaRoleType.TAIL:
+            self.name = f'runtime-{self.args.name}-tail'
+        if 'host' in self.args and 'port_ctrl' in self.args and 'ctrl_with_ipc' in self.args:
+            self.ctrl_addr, self.ctrl_with_ipc = Zmqlet.get_ctrl_address(self.args.host, self.args.port_ctrl,
+                                                                         self.args.ctrl_with_ipc)
+
+        if 'log_id' in self.args and 'log_config' in self.args:
+            self.logger = JinaLogger(self.name,
+                                     log_id=self.args.log_id,
+                                     log_config=self.args.log_config)
+        else:
+            self.logger = JinaLogger(self.name)
```

Comment:
can we rewrite this part and remove all unnecessary ones from `BaseRuntime` and move to subclasses?

Reply:
I think they are all needed here